docs/proposals/mutex-improvements.md

```diff
@@ -0,0 +1,96 @@
+
+`
+getHolderKey({"Namespace": "argo", Name: "deadlock-test-sn8p5"}, "deadlock-test-sn8p5") = "argo/deadlock-test-sn8p5/deadlock-test-sn8p5"
+
+-- "MutexStatus" After LockAcquired Call --
+items = ["argo", "deadlock-test-sn8p5", "deadlock-test-sn8p5"]
+holdingName = "deadlock-test-sn8p5"
+ms.Holding = [MutexHolding{Mutex: lockKey, Holder: holdingName}]
+
+getResourceKey("argo", "deadlock-test-sn8p5", "deadlock-test-sn8p5") = "argo/deadlock-test-sn8p5"
+`
+
+### A criticism of the current approach
+
+The current approach is heavily dependent on string formatting. This makes it incredibly difficult to understand functionality when there are no comments or documentation
+outlining the logic used. On top of this, it makes the "Find References" functionality on LSP powered editors near useless. My understanding of the function `getResourceKey` is that
+it seems to construct the holder key that is generated via `getHolderKey` through using processed strings generated via `LockAcquired`.
+I strongly believe this needs to be refactored to be less dependent on string formatting. I strongly believe that the key obtaining process should be far more obvious, the disconnect between key acquisition and release is inherently flawed and will only lead to further bugs down the line.
+
+### How should they work?
+
+We should not be generating information after it has already been generated. It is extremely important to only maintain a single source of truth,
+generating this information (the holder key) produces sources of truth at every generation point. Ideally we should be storing this information somewhere
+as we will be needing it eventually when a release call is made.
+
+### Solutions
+
+#### Store the holder key directly in the MutexStatus/SemaphoreStatus structs
+
+I may be wrong here but I don't see why it wouldn't be possible to store this information directly in the status structs.
+This seems to be the most simple way of ensuring that there is only a single source of truth. I didn't go with this solution in my existing PR because this would require changing the information MutexStatus/SemaphoreStatus held. But given this
+discussion is being opened in a proposal, it seems plausible to go with this solution instead.
+
+##### Advantages of solution #1
+
+* More sensible approach than including a new dependency (ConfigMap).
+* Relatively simple change.
+* Single source of truth
+
+##### Disadvantages of solution #1
+
+* Changing a field could be major change and might break projects in the ecosystem that rely on this behavior. This has been the behavior of mutexes ever since they have been implemented, it is high risk to change this in my opinion.
+
+#### Store holding/pending information in a config map
+
+We can store the holder keys inside a config map, on release, we refer to the holder keys inside this config map. There is an existing WIP PR for this, [here](https://github.com/argoproj/argo-workflows/pull/10009).
```

Comment:
Do you think there is a security concern in multi-tenant env?  If Controller updates the holder on Configmap, The controller should have cluster-level RBAC to update the config map across the namespace.